### PR TITLE
kube-apiserver: output flags in logical sections

### PR DESCRIPTION
**Pull request by sttts**
bildschirmfoto 2018-06-07 um 16 21 17

```release-note
Print kube-apiserver --help flag help in sections.
```

### CODE REVIEW

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -122,26 +122,35 @@ func NewServerRunOptions() *ServerRunOptions {
 	return &s
 }
 
-// AddFlags adds flags for a specific APIServer to the specified FlagSet
-func (s *ServerRunOptions) AddFlags(fs *pflag.FlagSet) {
+// Flags returns flags for a specific APIServer by section name
+func (s *ServerRunOptions) Flags() (fss map[string]*pflag.FlagSet) {
```
Comment (hzxuzhonghu) [inline cmd/kube-apiserver/app/options/options.go]: redundant fss

**Comment by hzxuzhonghu** [comment]
LGTM, it works fine.

The flags help info are well grouped instead of in alphabetical order.

```
The Kubernetes API server validates and configures data
for the api objects which include pods, services, replicationcontrollers, and
others. The API Server services REST operations and provides the frontend to the
cluster's shared state through which all other components interact.

Usage:
  kube-apiserver [flags]

Generic flags:
      --advertise-address ip                 The IP address on which to advertise the apiserver to members of the cluster. This address must be reachable by the rest of the cluster. If blank, the --bind-address will be used. If --bind-address is unspecified, the host's default interface will be used.
      --cors-allowed-origins strings         List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled.
      --external-hostname string             The hostname to use when generating externalized URLs for this master (e.g. Swagger API Docs).
      --feature-gates mapStringBool          A set of key=value pairs that describe feature gates for alpha/experimental features. Options are:
                APIListChunking=true|false (BETA - default=true)
                APIResponseCompression=true|false (ALPHA - default=false)
                AdvancedAuditing=true|false (BETA - default=true)
                AllAlpha=true|false (ALPHA - default=false)

```

**Comment by sttts** [comment]
It's a bit ugly that we cannot pass a uniform indention to cobra for all sections. Actually, I like the indention with the newline best:

```
    --foo-very-long-flag-so-that-not-space-is-left
         This does some thing cool, over many many many many many many many many
         many many many many many many many many lines.
```

This indention is only chosen implicitly by cobra if there is really not enough space. We would need to contribute something upstream to get this feature under control.

**Comment by sttts** [comment]
Fixed the order. Enforcing the "newline mode" now via a fake zzzzzzz flag.

**Comment by sttts** [comment]
This could need a follow-up to move the flags package into k8s.io/apiserver somewhere and re-use it e.g. in kube-controller-manager, kube-scheduler, cloud-controller-manager.

**Review by yue9944882 — COMMENTED** [review]
thanks! this sectional print makes perfect sense to me. but literally i'm concerned about it's package path. what if we introduce this utility to other cmds in the future? should they import sth like `k8s.io/cmd/kubeapiserver/...` then? or we could just put these to another common shared package. WDYT?  @sttts

**Comment by yue9944882** [comment]
@sttts not really sure. what about [https://github.com/kubernetes/utils](https://github.com/kubernetes/utils)?

**Comment by sttts** [comment]
@yue9944882 move it to k8s.io/apiserver/pkg/flags

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -122,26 +121,26 @@ func NewServerRunOptions() *ServerRunOptions {
...
-	s.Features.AddFlags(fs)
-	s.Authentication.AddFlags(fs)
-	s.Authorization.AddFlags(fs)
-	s.CloudProvider.AddFlags(fs)
-	s.StorageSerialization.AddFlags(fs)
-	s.APIEnablement.AddFlags(fs)
-	s.Admission.AddFlags(fs)
+	s.GenericServerRunOptions.AddUniversalFlags(fss.FlagSet("generic"))
+	s.Etcd.AddFlags(fss.FlagSet("etcd"))
+	s.SecureServing.AddFlags(fss.FlagSet("secure serving"))
```
Comment (stewart-yu) [inline cmd/kube-apiserver/app/options/options.go]: Using `secure-serving` may be better?

**Pull request merged**